### PR TITLE
fix!: improve interaction component typings

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Refactors modal interaction component typings which previously used full `ActionRow` types instead of partial types like the ones that are actually received in interactions.
`ModalInteractionData.[_]components` is now no longer a `List[ActionRow]` with mostly empty components, and instead just the raw data dict from the interaction, which avoids discarding currently unused data (like select values). This is unfortunately a breaking change, but necessary as the old typing was mostly wrong and could've resulted in issues with component parsing code due to missing fields.

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
    - *renamed `_components` to `components`, changed type*
    - *renamed `walk_components` to `walk_raw_components`*
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/interactions/modal.py [hunk 1]
```diff
@@ -134,19 +142,18 @@ class ModalInteractionData(Dict[str, Any]):
     ----------
     custom_id: :class:`str`
         The custom ID of the modal.
+    components: List[:class:`dict`]
+        The raw component data of the modal interaction.
+
+        .. versionadded:: 2.6
```
Comment (onerandomusername) [inline disnake/interactions/modal.py:148]: Should probably add something like 
> ...as provided by Discord. This is incomplete due to limitations by Discord.

I feel like the current wording people will miss that its incomplete

File: disnake/interactions/modal.py [hunk 2]
```diff
@@ -134,19 +142,18 @@ class ModalInteractionData(Dict[str, Any]):
...
     """
 
-    __slots__ = ("custom_id", "_components")
+    __slots__ = ("custom_id", "components")
 
     def __init__(self, *, data: ModalInteractionDataPayload):
         super().__init__(data)
         self.custom_id: str = data["custom_id"]
-        # this attribute is not meant to be used since it lacks most of the component data
```
Comment (onerandomusername) [inline disnake/interactions/modal.py:144]: we should keep a portion of this comment; the data is still incomplete.

File: disnake/interactions/modal.py [hunk 3]
```diff
@@ -101,22 +103,28 @@ def __init__(self, *, data: ModalInteractionPayload, state: ConnectionState):
             message = None
         self.message: Optional[Message] = message
 
-    def walk_components(self) -> Generator[ModalComponent, None, None]:
-        """Returns a generator that yields components from action rows one by one.
+    def walk_component_data(self) -> Generator[ModalInteractionComponentDataPayload, None, None]:
+        """
+        Returns a generator that yields raw component data from action rows one by one.
```
Comment (onerandomusername) [inline disnake/interactions/modal.py:108]: same comment here

**Review by onerandomusername — CHANGES_REQUESTED** [review]

File: disnake/interactions/modal.py (hunk 1 above)
Comment (shiftinv) [reply disnake/interactions/modal.py]: Good point, added in acca2ae01f712587e6d0c111b868801e49502c60.

File: disnake/interactions/modal.py (hunk 2 above)
Comment (shiftinv) [reply disnake/interactions/modal.py:144]: acca2ae01f712587e6d0c111b868801e49502c60

File: disnake/interactions/modal.py (hunk 3 above)
Comment (shiftinv) [reply disnake/interactions/modal.py]: acca2ae01f712587e6d0c111b868801e49502c60

**Review by onerandomusername — APPROVED** [review]
looks good

**Pull request merged**